extensions/amp-next-page/1.0/visibility-observer.js

```diff
@@ -178,4 +298,90 @@ export default class VisibilityObserver {
     installPositionObserverServiceForDoc(this.ampdoc_);
     return Services.positionObserverForDoc(this.ampdoc_.getHeadNode());
   }
+
+  /**
+   * Calculates the position of the element relative to the viewport
+   * based on the positions of the injected bottom and top sentinel elements
+   *
+   * @param {!VisibilityObserverEntry} entry
+   */
+  updateRelativePos(entry) {
+    const relativePos = entry.usesSentinel()
+      ? this.getRelativePosFromSentinel(entry)
+      : this.getRelativePosFromScroll(
+          /** @type {!VisibilityObserverHostEntry} */ (entry)
+        );
+    if (!relativePos) {
+      return;
+    }
+    entry.callback(devAssert(relativePos));
+  }
+
+  /**
+   * @param {!VisibilityObserverHostEntry} entry
+   * @return {?ViewportRelativePos}
+   */
+  getRelativePosFromScroll(entry) {
+    // Measure the position of the host page (edge case)
+    const {viewportHeight_: vh, lastScrollTop_: scroll} = this;
+    // Document height is the same as the distance from the top
+    // to the <amp-next-page> element
+    const {top: height} = entry.nextPageEl.getLayoutBox();
+    if (scroll < height - vh) {
+      return ViewportRelativePos.CONTAINS_VIEWPORT;
+    }
+    if (scroll < height && height <= vh && scroll <= 0) {
+      return ViewportRelativePos.INSIDE_VIEWPORT;
+    }
+    if (scroll < height) {
+      return this.isScrollingDown()
+        ? ViewportRelativePos.LEAVING_VIEWPORT
+        : ViewportRelativePos.ENTERING_VIEWPORT;
+    }
+    return ViewportRelativePos.OUTSIDE_VIEWPORT;
+  }
+
+  /**
+   * @param {!VisibilityObserverEntry} entry
+   * @return {?ViewportRelativePos}
+   */
+  getRelativePosFromSentinel(entry) {
```

Comment:
I'm confused about state flow.

Why is the concept of `sentinels` spilling to this level? Should this be localized only when you need a sentinel?

Reply:
Sentinel is actually the general case, the host page is a special case because we don't want to inject sentinel elements into the body of the document so we rely on the `amp-next-page` element position for calculations